vmm/src/device_manager.rs

```diff
@@ -1383,12 +1383,16 @@ impl DeviceManager {
                     .allocate_mmio_addresses(None, size as GuestUsize, Some(0x0020_0000))
                     .ok_or(DeviceManagerError::PmemRangeAllocation)?;
 
-                let (custom_flags, set_len) = if pmem_cfg.file.is_dir() {
+                let (mut custom_flags, set_len) = if pmem_cfg.file.is_dir() {
                     (O_TMPFILE, true)
                 } else {
                     (0, false)
                 };
 
+                if pmem_cfg.readonly {
+                    custom_flags |= O_RDONLY;
+                }
+
                 let file = OpenOptions::new()
                     .read(true)
                     .write(true)
```

Comment:
Why not just set .write(false) here?

Reply:
@eryugey thanks for the idea I incorporated it.